feat(ngcc): execute schematic-like migrations in ngcc

JoostK opened the pull request:
This PR contains several commits to introduce schematic-like migration in ngcc. More details can be found in each commit description.

JoostK committed:
feat(ngcc): migrate services that are missing `@Injectable()`

A class that is provided as Angular service is required to have an
`@Injectable()` decorator so that the compiler generates its injectable
definition for the runtime. Applications are automatically migrated
using the "missing-injectable" schematic, however libraries built for
older version of Angular may not yet satisfy this requirement.

This commit ports the "missing-injectable" schematic to a migration that
is ran when ngcc is processing a library. This ensures that any service
that is provided from an NgModule or Directive/Component will have an
`@Injectable()` decorator.

JoostK committed:
refactor(ivy): mark synthetic decorators explicitly

In ngcc's migration system, synthetic decorators can be injected into a
compilation to ensure that certain classes are compiled with Angular
logic, where the original library code did not include the necessary
decorators. Prior to this change, synthesized decorators would have a
fake AST structure as associated node and a made-up identifier. In
theory, this may introduce issues downstream:

1) a decorator's node is used for diagnostics, so it must have position
information. Having fake AST nodes without a position is therefore a
problem. Note that this is currently not a problem in practice, as
injected synthesized decorators would not produce any diagnostics.

2) the decorator's identifier should refer to an imported symbol.
Therefore, it is required that the symbol is actually imported.
Moreover, bundle formats such as UMD and CommonJS use namespaces for
imports, so a bare `ts.Identifier` would not be suitable to use as
identifier. This was also not a problem in practice, as the identifier
is only used in the `setClassMetadata` generated code, which is omitted
for synthetically injected decorators.

To remedy these potential issues, this commit makes a decorator's
identifier optional and switches its node over from a fake AST structure
to the class' name.

JoostK committed:
refactor(ngcc): rework undecorated parent migration

Previously, the (currently disabled) undecorated parent migration in
ngcc would produce errors when a base class could not be determined
statically or when a class extends from a class in another package. This
is not ideal, as it would cause the library to fail compilation without
a workaround, whereas those problems are not guaranteed to cause issues.

Additionally, inheritance chains were not handled. This commit reworks
the migration to address these limitations.

JoostK committed:
feat(ngcc): enable migrations to apply schematics to libraries

When upgrading an Angular application to a new version using the Angular
CLI, built-in schematics are being run to update user code from
deprecated patterns to the new way of working. For libraries that have
been built for older versions of Angular however, such schematics have
not been executed which means that deprecated code patterns may still be
present, potentially resulting in incorrect behavior.

Some of the logic of schematics has been ported over to ngcc migrations,
which are automatically run on libraries. These migrations achieve the
same goal of the regular schematics, but operating on published library
sources instead of used code.

trotyl commented on packages/compiler-cli/ngcc/src/migrations/missing_injectable_migration.ts:
```diff
@@ -0,0 +1,184 @@
+      return;
+    }
+    if (provider.has('useExisting')) {
+      migrateProviderClass(provider.get('useExisting') !, host);
```
A class referenced by `useExisting` can be pure token (as if not a class at all), not implementation, which does not require an `@Injectable()` even in Ivy.

devversion replied:
That is a very good point. Thanks for raising this. I'll be looking into this briefly tomorrow, but I'm quite sure that this is currently just an oversight in the `ng update` migration (which the ngcc migration basically follows)

JoostK replied:
@trotyl That is a very good point indeed! This appears to have been an oversight in the "missing-injectable" schematic, for which @devversion will open a separate fix. I'm fixing the ngcc migration in this PR.

JoostK commented:
Note to self: port #33286 over to this PR as well

JoostK committed:
fix(ngcc): prevent reflected decorators from being clobbered

ngcc has an internal cache of computed decorator information for
reflected classes, which could previously be mutated by consumers of the
reflection host. With the ability to inject synthesized decorators, such
decorators would inadvertently be added into the array of decorators
that was owned by the internal cache of the reflection host, incorrectly
resulting in synthesized decorators to be considered real decorators on
a class. This commit fixes the issue by cloning the cached array before
returning it.

JoostK commented:
After having tested within the ngcc-validation project in https://github.com/angular/ngcc-validation/pull/439 (thanks @gkalpak!) two issues were identifier:

1. synthetic decorators did get emitted as metadata, which should not be the case. This has been addressed by preventing the internal decorators cache of `Esm2015ReflectionHost` from being clobbered with outside changes, such as injecting synthetic decorators. With this change, synthetic decorators no longer leak into the real decorators so won't be emitted as metadata.
1. Abstract directives could be migrated despite already having a `@Directive` decorator, causing an error. This error was addressed by updating `hasDirectiveDirector` to also consider `MetadataReader.isAbstractDirective`, and tests have been adjusted to assert that no errors occurred. Note that this change is obsolete once once #33131 has landed.

I am fairly confident that the migrations now operate as they should. The two packages that regressed in the test run are now building successfully for me locally.

gkalpak commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -347,7 +347,12 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
   /** Gets all decorators of the given class symbol. */
   getDecoratorsOfSymbol(symbol: NgccClassSymbol): Decorator[]|null {
     const {classDecorators} = this.acquireDecoratorInfo(symbol);
```
Consider adding a comment on `decoratroCache` and `acquireDecoratorInfo()` to mention that synthetic decorators are not included in the cached/returned `DecoratorInfo`.

gkalpak commented on packages/compiler-cli/ngcc/test/migrations/undecorated_parent_migration_spec.ts:
```diff
@@ -78,20 +80,87 @@ runInEachFileSystem(() => {
+         expect(decorator.name).toEqual('Directive');
+         expect(decorator.identifier).toBeNull('The decorator must be synthesized');
+         expect(decorator.import).toEqual({from: '@angular/core', name: 'Directive'});
+         expect(decorator.args !.length).toEqual(0);
```
Could you add some assertions about what happens to `IntermediateClass`.

gkalpak commented on packages/compiler-cli/ngcc/src/migrations/missing_injectable_migration.ts:
```diff
@@ -0,0 +1,183 @@
+ * `Directive.providers`/`Component.viewProviders` also have an `@Injectable()` decorator. This
+ * decorator is now mandatory, as otherwise the compiler would not compile an injectable definition
+ * for the service. This is unlike View Engine, where having just an unrelated decorator may have
+ * been sufficient for the service to become injectable.
```
This paragraph confuses me a bit. By reading it, I thought that we need the `@Injectable` decorator no matter what. By looking at the code, it seems that `@Injectable` is needed only if the class does not have `@Directive`, `@Component` or `@Pipe`.

gkalpak reviewed (approved):
Great stuff 💯 

<sub>I wish there was some way to share logic with the schematics migrations, but I guess the "infrastructure" is too different for that 🤷‍♂</sub>

JoostK commented on packages/compiler-cli/src/ngtsc/reflection/src/host.ts:
```diff
@@ -32,7 +33,8 @@ export interface Decorator {
   /**
-   * TypeScript reference to the decorator itself.
+   * TypeScript reference to the decorator itself, or the node for which the decorator was
+   * synthesized in ngcc.
```
`Decorator.node` is primarily used for producing diagnostics (for its location) so it would be quite cumbersome if this could be `null`. For synthesized decorators we don't expect diagnostics to be created, so from that perspective this shouldn't matter much, however I liked using the migrated class' identifier as a fallback.

JoostK commented on packages/compiler-cli/ngcc/test/analysis/decoration_analyzer_spec.ts:
```diff
@@ -199,10 +199,10 @@ runInEachFileSystem(() => {
-            'migration2:MyDirective',
             'migration1:MyOtherComponent',
+            'migration2:MyComponent',
+            'migration2:MyDirective',
```
The main reason for changing the ordering is my sanity, basically. Migrations are ran on classes and typically migrate _other_ classes that are referred to somehow, so if migrations were to be executed intertwined there would be an ill-defined ordering of how multiple migrations interact on each other, as the ordering of classes and their references is not at all stable.

This is also why `UndecoratedParentMigration` precedes `MissingInjectableMigration` in the list of migrations, as we don't want to introduce an `@Injectable()` for something where an `@Directive` would be more appropriate.

JoostK committed:
fixup! fix(ngcc): prevent reflected decorators from being clobbered

JoostK committed:
fixup! feat(ngcc): migrate services that are missing `@Injectable()`

JoostK committed:
fixup! refactor(ngcc): rework undecorated parent migration

JoostK commented:
Closing as these commits are part of #33362 to be landed at once.
